include/deal.II/base/work_stream.h

```diff
@@ -874,14 +1019,352 @@ namespace WorkStream
 
 
+
+#  ifdef DEAL_II_WITH_TASKFLOW
+    namespace taskflow_v4
+    {
+      template <typename CopyData>
+      class Chunk
+      {
+      public:
+        Chunk(const unsigned int count, const CopyData &copy_data)
+          : copy_datas(count, copy_data)
+        {}
+        std::vector<CopyData> copy_datas;
+      };
+
+
+
+      /**
+       * WorkStream without colored iterators.
+       *
+       * Work in chunks of size @p chunk_size (only if we have enough items, otherwise 1) and
+       * use a thread-local scratch object
+       */
+      template <typename Worker,
+                typename Copier,
+                typename Iterator,
+                typename ScratchData,
+                typename CopyData>
+      void
+      run(const Iterator &                         begin,
+          const typename identity<Iterator>::type &end,
+          Worker                                   worker,
+          Copier                                   copier,
+          const ScratchData &                      sample_scratch_data,
+          const CopyData &                         sample_copy_data,
+          const unsigned int                       chunk_size = 8)
+      {
+        if (MultithreadInfo::n_threads() == 1)
+          {
+            // need to copy the sample since it is marked const
+            ScratchData scratch_data = sample_scratch_data;
+            CopyData    copy_data    = sample_copy_data; // NOLINT
+
+            for (Iterator i = begin; i != end; ++i)
+              {
+                // need to check if the function is not the zero function. To
+                // check zero-ness, create a C++ function out of it and check
+                // that
+                if (static_cast<const std::function<void(
+                      const Iterator &, ScratchData &, CopyData &)> &>(worker))
+                  worker(i, scratch_data, copy_data);
+                if (static_cast<const std::function<void(const CopyData &)> &>(
+                      copier))
+                  copier(copy_data);
+              }
+
+            return;
+          }
+
+        tf::Executor &executor = MultithreadInfo::get_taskflow_executor();
+        tf::Taskflow  taskflow;
+
+        tf::Task last_copier = taskflow.placeholder();
+
+        Threads::ThreadLocalStorage<std::unique_ptr<ScratchData>>
+          thread_local_scratch;
+
+        std::vector<std::unique_ptr<Chunk<CopyData>>> chunks;
+
+        unsigned int idx             = 0;
+        unsigned int remaining_items = std::distance(begin, end);
+
+        const unsigned int real_chunk_size =
+          (remaining_items / chunk_size < 3 * MultithreadInfo::n_threads()) ?
+            1 :
+            chunk_size;
+
+        Iterator it = begin;
+        while (it != end)
+          {
+            unsigned int count  = std::min(remaining_items, real_chunk_size);
+            Iterator     middle = it;
+            std::advance(middle, count);
+
+            chunks.emplace_back();
```

Comment:
You're allocating a single element here. Can you document what this is for? I'm asking because a few lines further down, you're accessing `chunks[idx]` but it's not clear to me at that point where `chunks` is correctly sized.